tools/functional-tester/etcd-tester/runner.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"strings"
+	"sync"
+
+	rl "github.com/coreos/etcd/tools/functional-tester/etcd-runner/lib"
+)
+
+var (
+	clientTimeout = 5 // seconds
+)
+
+type Runner interface {
+	// Run runs runner with current round #
+	Run(round int)
+	// Cancel() waits runner to finish
+	Cancel()
+	// Check checks if there are any errors occured during Run()
+	Check() error
+}
+
+type noRunner struct {
+}
+
+func (n *noRunner) Run(round int) {
+	// Do nothing
+}
+
+func (n *noRunner) Cancel() {
+	// Do nothing
+}
+
+func (n *noRunner) Check() error {
+	return nil
+}
+
+type compositeRunner struct {
+	runners []Runner
+}
+
+func (cr *compositeRunner) Run(round int) {
+	for i := range cr.runners {
+		go cr.runners[i].Run(round)
+	}
+}
+
+func (cr *compositeRunner) Cancel() {
+	var wg sync.WaitGroup
+	wg.Add(len(cr.runners))
+	for i := range cr.runners {
+		go func(idx int) {
+			defer wg.Done()
+			cr.runners[idx].Cancel()
+		}(i)
+	}
+	wg.Wait()
+}
+
+func (cr *compositeRunner) Check() error {
+	for i := range cr.runners {
+		if err := cr.runners[i].Check(); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type electionRunner struct {
+	cf    *rl.EtcdRunnerConfig
+	runWg sync.WaitGroup
+	errch chan error
+}
+
+func (er *electionRunner) Run(round int) {
+	er.errch = make(chan error, 1)
+	er.runWg.Add(1)
+	go func() {
+		defer er.runWg.Done()
+		er.errch <- rl.RunElection(er.cf)
+	}()
+}
+
+func (er *electionRunner) Cancel() {
+	er.runWg.Wait()
+}
+
+func (er *electionRunner) Check() error {
+	return <-er.errch
+}
+
+type lockRacerRunner struct {
+	cf    *rl.EtcdRunnerConfig
+	runWg sync.WaitGroup
+	errch chan error
+}
+
+func (lr *lockRacerRunner) Run(round int) {
+	lr.errch = make(chan error, 1)
+	lr.runWg.Add(1)
+	go func() {
+		defer lr.runWg.Done()
+		lr.errch <- rl.RunRacer(lr.cf)
+	}()
+}
+
+func (lr *lockRacerRunner) Cancel() {
+	lr.runWg.Wait()
+}
+
+func (lr *lockRacerRunner) Check() error {
+	return <-lr.errch
+}
+
+type watchRunner struct {
+	cf    *rl.EtcdRunnerConfig
+	runWg sync.WaitGroup
+	errch chan error
+}
+
+func (wr *watchRunner) Run(round int) {
+	wr.errch = make(chan error, 1)
+	wr.runWg.Add(1)
+	go func() {
+		defer wr.runWg.Done()
+		wr.errch <- rl.RunWatchOnce(wr.cf, round)
+	}()
+}
+
+func (wr *watchRunner) Cancel() {
+	wr.runWg.Wait()
```

Comment:
This doesn't really cancel, it only waits until everything completes. The tester should be able to terminate stressers without waiting a long time.